Turn off netty client DNS connect retry.

<!--- Provide a general summary of the issue in the Title above -->
Turn off netty client DNS connect retry.

[StackOverflow](https://stackoverflow.com/questions/69218709/how-can-i-turn-off-netty-client-dns-connect-retry?noredirect=1#comment122351702_69218709)
<!--- /!\ Make sure to follow the Contribution Guidelines and notably for security issues:
https://github.com/reactor/reactor-netty/blob/main/CONTRIBUTING.md
https://tanzu.vmware.com/security
-->
<!--- /!\ Questions should be asked on [Gitter](https://gitter.im/reactor/reactor-netty) or [StackOverflow](https://stackoverflow.com/questions/tagged/reactor-netty). -->

## Motivation
<!--- A clear and concise description of what the problem is. Ex. I'm always frustrated when [...] -->

I'm using Netty-httpClient for Spring webClient.

When I connect  `www.naver.com` and set the host like this.

```
127.0.0.1 www.naver.com
125.209.222.142 www.naver.com
```
And run httpClient.get() result like this.

```
 DEBUG r.n.r.PooledConnectionProvider - [id:83c8069f] Created a new pooled channel, now: 0 active connections, 0 inactive connections and 0 pending acquire requests.
 DEBUG r.n.t.SslProvider - [id:83c8069f] SSL enabled using engine sun.security.ssl.SSLEngineImpl@5412b204 and SNI www.naver.com:443
 DEBUG i.n.b.AbstractByteBuf - -Dio.netty.buffer.checkAccessible: true
 DEBUG i.n.b.AbstractByteBuf - -Dio.netty.buffer.checkBounds: true
 DEBUG i.n.u.ResourceLeakDetectorFactory - Loaded default ResourceLeakDetector: io.netty.util.ResourceLeakDetector@7ec1e7a5
 DEBUG r.n.t.TransportConfig - [id:83c8069f] Initialized pipeline DefaultChannelPipeline{(reactor.left.sslHandler = io.netty.handler.ssl.SslHandler), (reactor.left.loggingHandler = reactor.netty.transport.logging.ReactorNettyLoggingHandler), (reactor.left.sslReader = reactor.netty.tcp.SslProvider$SslReadHandler), (reactor.left.httpCodec = io.netty.handler.codec.http.HttpClientCodec), (reactor.right.reactiveBridge = reactor.netty.channel.ChannelOperationsHandler)}
 INFO  reactor - [id:83c8069f] REGISTERED
 DEBUG r.n.t.TransportConnector - [id:83c8069f] Connecting to [www.naver.com/127.0.0.1:443].
 INFO  reactor - [id:83c8069f] CONNECT: www.naver.com/127.0.0.1:443
 INFO  reactor - [id:83c8069f] CLOSE
 DEBUG r.n.t.TransportConnector - [id:83c8069f] Connect attempt to [www.naver.com/127.0.0.1:443] failed.
io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: no further information: www.naver.com/127.0.0.1:443
Caused by: java.net.ConnectException: Connection refused: no further information
    at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
    at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:716)
    at io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:330)
    at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:334)
    at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:707)
    at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:655)
    at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:581)
    at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
    at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
    at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
    at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
    at java.lang.Thread.run(Thread.java:748)
 DEBUG r.n.r.PooledConnectionProvider - [id:0e3c272c] Created a new pooled channel, now: 0 active connections, 0 inactive connections and 0 pending acquire requests.
 DEBUG r.n.t.SslProvider - [id:0e3c272c] SSL enabled using engine sun.security.ssl.SSLEngineImpl@47b241b and SNI www.naver.com:443
 DEBUG r.n.t.TransportConfig - [id:0e3c272c] Initialized pipeline DefaultChannelPipeline{(reactor.left.sslHandler = io.netty.handler.ssl.SslHandler), (reactor.left.loggingHandler = reactor.netty.transport.logging.ReactorNettyLoggingHandler), (reactor.left.sslReader = reactor.netty.tcp.SslProvider$SslReadHandler), (reactor.left.httpCodec = io.netty.handler.codec.http.HttpClientCodec), (reactor.right.reactiveBridge = reactor.netty.channel.ChannelOperationsHandler)}
 INFO  reactor - [id:0e3c272c] REGISTERED
 INFO  reactor - [id:83c8069f] UNREGISTERED
 DEBUG r.n.t.TransportConnector - [id:0e3c272c] Connecting to [www.naver.com/125.209.222.142:443].
 INFO  reactor - [id:0e3c272c] CONNECT: www.naver.com/125.209.222.142:443
 DEBUG r.n.r.DefaultPooledConnectionProvider - [id:0e3c272c, L:/192.168.55.77:52910 - R:www.naver.com/125.209.222.142:443] Registering pool release on close event for channel
 DEBUG r.n.r.PooledConnectionProvider - [id:0e3c272c, L:/192.168.55.77:52910 - R:www.naver.com/125.209.222.142:443] Channel connected, now: 1 active connections, 0 inactive connections and 0 pending acquire requests.
 DEBUG i.n.u.Recycler - -Dio.netty.recycler.maxCapacityPerThread: 4096
 DEBUG i.n.u.Recycler - -Dio.netty.recycler.maxSharedCapacityFactor: 2
 DEBUG i.n.u.Recycler - -Dio.netty.recycler.linkCapacity: 16
 DEBUG i.n.u.Recycler - -Dio.netty.recycler.ratio: 8
 DEBUG i.n.u.Recycler - -Dio.netty.recycler.delayedQueue.ratio: 8
 INFO  reactor - [id:0e3c272c, L:/192.168.55.77:52910 - R:www.naver.com/125.209.222.142:443] ACTIVE
```

see that. I don't want connect twice and don't spend more time to connect.
just fail if can not make connect.


## Desired solution
<!--- A clear and concise description of what you want to happen. -->
Configure DNS resolve timeout and requestTimeout include this if possible. 


## Considered alternatives
<!--- A clear and concise description of any alternative solutions or features you've considered. -->

## Additional context
<!--- Add any other context or screenshots about the feature request here. -->

--- a/docs/asciidoc/http-client.adoc
+++ b/docs/asciidoc/http-client.adoc
@@ -492,6 +492,8 @@ This configuration is applicable for `DnsNameResolver#resolveAll(String)`. By de
 how much data the resolver can read per response. By default, this setting is enabled.
 | `disableRecursionDesired` | Specifies whether this resolver has to send a DNS query with the recursion desired (RD) flag set.
 By default, this setting is enabled.
+| `hostsFileEntriesResolver` | Sets a custom {nettyjavadoc}/io/netty/resolver/HostsFileEntriesResolver.html[`HostsFileEntriesResolver`] to be used for hosts file entries.
+Default: {nettyjavadoc}/io/netty/resolver/DefaultHostsFileEntriesResolver.html[`DefaultHostsFileEntriesResolver`].
 | `maxPayloadSize` | Sets the capacity of the datagram packet buffer (in bytes). Default: 4096.
 | `maxQueriesPerResolve` | Sets the maximum allowed number of DNS queries to send when resolving a host name. Default: 16.
 | `ndots` | Sets the number of dots that must appear in a name before an initial absolute query is made.

--- a/docs/asciidoc/tcp-client.adoc
+++ b/docs/asciidoc/tcp-client.adoc
@@ -408,6 +408,8 @@ This configuration is applicable for `DnsNameResolver#resolveAll(String)`. By de
  how much data the resolver can read per response. By default, this setting is enabled.
 | `disableRecursionDesired` | Specifies whether this resolver has to send a DNS query with the recursion desired (RD) flag set.
  By default, this setting is enabled.
+| `hostsFileEntriesResolver` | Sets a custom {nettyjavadoc}/io/netty/resolver/HostsFileEntriesResolver.html[`HostsFileEntriesResolver`] to be used for hosts file entries.
+Default: {nettyjavadoc}/io/netty/resolver/DefaultHostsFileEntriesResolver.html[`DefaultHostsFileEntriesResolver`].
 | `maxPayloadSize` | Sets the capacity of the datagram packet buffer (in bytes). Default: 4096.
 | `maxQueriesPerResolve` | Sets the maximum allowed number of DNS queries to send when resolving a host name. Default: 16.
 | `ndots` | Sets the number of dots that must appear in a name before an initial absolute query is made.

--- a/reactor-netty-core/src/main/java/reactor/netty/transport/NameResolverProvider.java
+++ b/reactor-netty-core/src/main/java/reactor/netty/transport/NameResolverProvider.java
@@ -20,6 +20,8 @@ import io.netty.channel.socket.DatagramChannel;
 import io.netty.channel.socket.SocketChannel;
 import io.netty.handler.logging.LogLevel;
 import io.netty.resolver.AddressResolverGroup;
+import io.netty.resolver.DefaultHostsFileEntriesResolver;
+import io.netty.resolver.HostsFileEntriesResolver;
 import io.netty.resolver.ResolvedAddressTypes;
 import io.netty.resolver.dns.DnsAddressResolverGroup;
 import io.netty.resolver.dns.DnsNameResolver;
@@ -92,7 +94,7 @@ public final class NameResolverProvider {
 		 * If {@code false}, the resolver notifies the returned {@link Future} when
 		 * all possible address types are complete.
 		 * This configuration is applicable for {@link DnsNameResolver#resolveAll(String)}.
-		 * By default this is enabled.
+		 * By default, this is enabled.
 		 *
 		 * @param enable {@code true} to enable, {@code false} to disable.
 		 * @return {@code this}
@@ -101,7 +103,7 @@ public final class NameResolverProvider {
 
 		/**
 		 * Disables the automatic inclusion of an optional record that tries to hint the remote DNS server about
-		 * how much data the resolver can read per response. By default this is enabled.
+		 * how much data the resolver can read per response. By default, this is enabled.
 		 *
 		 * @param disable true if an optional record is not included
 		 * @return {@code this}
@@ -110,12 +112,21 @@ public final class NameResolverProvider {
 
 		/**
 		 * Specifies whether this resolver has to send a DNS query with the recursion desired (RD) flag set.
-		 * By default this is enabled.
+		 * By default, this is enabled.
 		 *
 		 * @param disable true if RD flag is not set
 		 * @return {@code this}
 		 */
 		NameResolverSpec disableRecursionDesired(boolean disable);
+
+		/**
+		 * Specifies a custom {@link HostsFileEntriesResolver} to be used for hosts file entries.
+		 * Default to {@link DefaultHostsFileEntriesResolver}.
+		 *
+		 * @param hostsFileEntriesResolver the {@link HostsFileEntriesResolver} to be used for hosts file entries
+		 * @return {@code this}
+		 */
+		NameResolverSpec hostsFileEntriesResolver(HostsFileEntriesResolver hostsFileEntriesResolver);
 
 		/**
 		 * Sets the capacity of the datagram packet buffer (in bytes).
@@ -206,7 +217,7 @@ public final class NameResolverProvider {
 
 		/**
 		 * Sets the list of search domains of the resolver.
-		 * By default the effective search domain list will be populated using
+		 * By default, the effective search domain list will be populated using
 		 * the system DNS search domains.
 		 *
 		 * @param searchDomains the search domains
@@ -259,6 +270,16 @@ public final class NameResolverProvider {
 	 */
 	public Duration cacheNegativeTimeToLive() {
 		return cacheNegativeTimeToLive;
+	}
+
+	/**
+	 * Returns the configured custom {@link HostsFileEntriesResolver} to be used for hosts file entries or null.
+	 *
+	 * @return the configured custom {@link HostsFileEntriesResolver} to be used for hosts file entries or null
+	 */
+	@Nullable
+	public HostsFileEntriesResolver hostsFileEntriesResolver() {
+		return hostsFileEntriesResolver;
 	}
 
 	/**
@@ -442,6 +463,9 @@ public final class NameResolverProvider {
 				.eventLoop(group.next())
 				.channelFactory(() -> loop.onChannel(DatagramChannel.class, group))
 				.socketChannelFactory(() -> loop.onChannel(SocketChannel.class, group));
+		if (hostsFileEntriesResolver != null) {
+			builder.hostsFileEntriesResolver(hostsFileEntriesResolver);
+		}
 		if (loggingFactory != null) {
 			builder.dnsQueryLifecycleObserverFactory(loggingFactory);
 		}
@@ -460,6 +484,7 @@ public final class NameResolverProvider {
 	final boolean completeOncePreferredResolved;
 	final boolean disableRecursionDesired;
 	final boolean disableOptionalRecord;
+	final HostsFileEntriesResolver hostsFileEntriesResolver;
 	final DnsQueryLifecycleObserverFactory loggingFactory;
 	final LoopResources loopResources;
 	final int maxPayloadSize;
@@ -478,6 +503,7 @@ public final class NameResolverProvider {
 		this.completeOncePreferredResolved = build.completeOncePreferredResolved;
 		this.disableOptionalRecord = build.disableOptionalRecord;
 		this.disableRecursionDesired = build.disableRecursionDesired;
+		this.hostsFileEntriesResolver = build.hostsFileEntriesResolver;
 		this.loggingFactory = build.loggingFactory;
 		this.loopResources = build.loopResources;
 		this.maxPayloadSize = build.maxPayloadSize;
@@ -506,6 +532,7 @@ public final class NameResolverProvider {
 		boolean completeOncePreferredResolved = DEFAULT_COMPLETE_ONCE_PREFERRED_RESOLVED;
 		boolean disableOptionalRecord;
 		boolean disableRecursionDesired;
+		HostsFileEntriesResolver hostsFileEntriesResolver;
 		DnsQueryLifecycleObserverFactory loggingFactory;
 		LoopResources loopResources;
 		int maxPayloadSize = DEFAULT_MAX_PAYLOAD_SIZE;
@@ -550,6 +577,12 @@ public final class NameResolverProvider {
 		@Override
 		public NameResolverSpec disableRecursionDesired(boolean disable) {
 			this.disableRecursionDesired = disable;
+			return this;
+		}
+
+		@Override
+		public NameResolverSpec hostsFileEntriesResolver(HostsFileEntriesResolver hostsFileEntriesResolver) {
+			this.hostsFileEntriesResolver = Objects.requireNonNull(hostsFileEntriesResolver);
 			return this;
 		}
 

--- a/reactor-netty-core/src/test/java/reactor/netty/transport/ClientTransportTest.java
+++ b/reactor-netty-core/src/test/java/reactor/netty/transport/ClientTransportTest.java
@@ -23,6 +23,7 @@ import io.netty.resolver.AddressResolverGroup;
 import io.netty.resolver.HostsFileEntriesProvider;
 import io.netty.resolver.NoopAddressResolverGroup;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.condition.DisabledOnOs;
 import reactor.core.publisher.Mono;
 import reactor.netty.Connection;
 import reactor.netty.channel.ChannelMetricsRecorder;
@@ -46,6 +47,7 @@ import java.util.stream.Collectors;
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
+import static org.junit.jupiter.api.condition.OS.WINDOWS;
 
 /**
  * @author Violeta Georgieva
@@ -234,12 +236,20 @@ class ClientTransportTest {
 		);
 	}
 
+	/**
+	 * On Windows OS it is rare to have hosts file
+	 */
 	@Test
+	@DisabledOnOs(WINDOWS)
 	void testDefaultHostsFileEntriesResolver() throws Exception {
 		doTestHostsFileEntriesResolver(false);
 	}
 
+	/**
+	 * On Windows OS it is rare to have hosts file
+	 */
 	@Test
+	@DisabledOnOs(WINDOWS)
 	void testCustomHostsFileEntriesResolver() throws Exception {
 		doTestHostsFileEntriesResolver(true);
 	}
@@ -280,7 +290,7 @@ class ClientTransportTest {
 
 			assertThat(latch.await(5, TimeUnit.SECONDS)).as("latch await").isTrue();
 
-			assertThat(resolved).isNotNull();
+			assertThat(resolved.get()).isNotNull();
 			if (customResolver) {
 				assertThat(resolved.get()).hasSize(1);
 				assertThat(resolved.get().get(0)).isEqualTo(addresses.get(0));

--- a/reactor-netty-core/src/test/java/reactor/netty/transport/ClientTransportTest.java
+++ b/reactor-netty-core/src/test/java/reactor/netty/transport/ClientTransportTest.java
@@ -20,6 +20,7 @@ import io.netty.channel.embedded.EmbeddedChannel;
 import io.netty.channel.nio.NioEventLoopGroup;
 import io.netty.handler.logging.LoggingHandler;
 import io.netty.resolver.AddressResolverGroup;
+import io.netty.resolver.HostsFileEntriesProvider;
 import io.netty.resolver.NoopAddressResolverGroup;
 import org.junit.jupiter.api.Test;
 import reactor.core.publisher.Mono;
@@ -29,15 +30,19 @@ import reactor.netty.resources.ConnectionProvider;
 import reactor.netty.resources.LoopResources;
 import reactor.util.annotation.Nullable;
 
+import java.net.InetAddress;
 import java.net.InetSocketAddress;
 import java.net.SocketAddress;
 import java.time.Duration;
 import java.util.Collections;
+import java.util.List;
 import java.util.Map;
 import java.util.Properties;
+import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Supplier;
+import java.util.stream.Collectors;
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
@@ -227,6 +232,73 @@ class ClientTransportTest {
 		return new TestClientTransport(Mono.empty(),
 				new TestClientTransportConfig(provider, Collections.emptyMap(), () -> null)
 		);
+	}
+
+	@Test
+	void testDefaultHostsFileEntriesResolver() throws Exception {
+		doTestHostsFileEntriesResolver(false);
+	}
+
+	@Test
+	void testCustomHostsFileEntriesResolver() throws Exception {
+		doTestHostsFileEntriesResolver(true);
+	}
+
+	@SuppressWarnings("unchecked")
+	private void doTestHostsFileEntriesResolver(boolean customResolver) throws Exception {
+		LoopResources loop1 = LoopResources.create("test", 1, true);
+		NioEventLoopGroup loop2 = new NioEventLoopGroup(1);
+		ConnectionProvider provider = ConnectionProvider.create("test");
+		TestClientTransportConfig config =
+				new TestClientTransportConfig(provider, Collections.emptyMap(), () -> null);
+		HostsFileEntriesProvider hostsFileEntriesProvider = HostsFileEntriesProvider.parser().parseSilently();
+		List<InetAddress> addresses = hostsFileEntriesProvider.ipv4Entries().get("localhost");
+		try {
+			config.loopResources = loop1;
+			if (customResolver) {
+				NameResolverProvider nameResolverProvider = NameResolverProvider.builder()
+						.hostsFileEntriesResolver((inetHost, resolvedAddressTypes) ->
+							// Returns only the first entry from the hosts file
+							addresses != null && !addresses.isEmpty() ? addresses.get(0) : null)
+						.build();
+				config.defaultResolver.set(nameResolverProvider.newNameResolverGroup(config.loopResources, true));
+			}
+
+			AtomicReference<List<InetAddress>> resolved = new AtomicReference<>();
+			CountDownLatch latch = new CountDownLatch(1);
+			config.resolverInternal()
+			      .getResolver(loop2.next())
+			      .resolveAll(InetSocketAddress.createUnresolved("localhost", 443))
+			      .addListener(f -> {
+			          resolved.set(
+			              ((List<InetSocketAddress>) f.getNow())
+			                      .stream()
+			                      .map(InetSocketAddress::getAddress)
+			                      .collect(Collectors.toList()));
+			          latch.countDown();
+			      });
+
+			assertThat(latch.await(5, TimeUnit.SECONDS)).as("latch await").isTrue();
+
+			assertThat(resolved).isNotNull();
+			if (customResolver) {
+				assertThat(resolved.get()).hasSize(1);
+				assertThat(resolved.get().get(0)).isEqualTo(addresses.get(0));
+			}
+			else {
+				assertThat(resolved.get()).hasSize(addresses.size());
+				assertThat(resolved.get()).isEqualTo(addresses);
+			}
+		}
+		finally {
+			config.defaultResolver.get().close();
+			loop1.disposeLater()
+			     .block(Duration.ofSeconds(10));
+			provider.disposeLater()
+			        .block(Duration.ofSeconds(10));
+			loop2.shutdownGracefully()
+			     .get(10, TimeUnit.SECONDS);
+		}
 	}
 
 	static final class TestClientTransport extends ClientTransport<TestClientTransport, TestClientTransportConfig> {

--- a/reactor-netty-core/src/test/java/reactor/netty/transport/NameResolverProviderTest.java
+++ b/reactor-netty-core/src/test/java/reactor/netty/transport/NameResolverProviderTest.java
@@ -134,6 +134,20 @@ class NameResolverProviderTest {
 	}
 
 	@Test
+	void hostsFileEntriesResolver() {
+		assertThat(builder.build().hostsFileEntriesResolver).isNull();
+
+		builder.hostsFileEntriesResolver((inetHost, resolvedAddressTypes) -> null);
+		assertThat(builder.build().hostsFileEntriesResolver).isNotNull();
+	}
+
+	@Test
+	void hostsFileEntriesResolverBadValues() {
+		assertThatExceptionOfType(NullPointerException.class)
+				.isThrownBy(() -> builder.hostsFileEntriesResolver(null));
+	}
+
+	@Test
 	void maxPayloadSize() {
 		assertThat(builder.build().maxPayloadSize()).isEqualTo(DEFAULT_MAX_PAYLOAD_SIZE);
 

